server/src/bundle-validation.ts

```diff
@@ -0,0 +1,198 @@
+import { Connection, Diagnostic, DiagnosticSeverity } from 'vscode-languageserver/node'
+import { BookBundle, BundleItem } from './book-bundle'
+import { calculateElementPositions, expect, generateDiagnostic } from './utils'
+
+enum DiagnosticSource {
+  ImagePath = 'Image validation',
+  Link = 'Link validation'
+}
+
+export interface BundleValidationRequest {
+  causeUri: string
+}
+
+export class BundleValidationQueue {
+  private queue: BundleItem[] = []
+  private timer: NodeJS.Immediate | undefined
+  errorEncountered: Error | undefined
+
+  constructor(private readonly bundle: BookBundle, private readonly connection: Connection) {}
+
+  private clearQueue(): void {
+    this.queue = []
+  }
+
+  addRequest(request?: BundleValidationRequest): void {
+    this.clearQueue()
+    if (request != null) {
+      const priorityItem = expect(this.bundle.bundleItemFromUri(request.causeUri), 'caller must verify uri resides in this bundle')
+      if (priorityItem.type === 'collections' || priorityItem.type === 'modules') {
+        this.queue.push(priorityItem)
+      }
+    }
+    this.queue.push(...this.bundle.collectionItems())
+    this.queue.push(...this.bundle.moduleItems())
+    this.trigger()
+  }
+
+  private async processQueue(): Promise<void> {
+    // This is slower than pop, but we shouldn't ever have more than a couple hundred items.
+    // It's still sub-ms total to shift all the items in total
+    const item = this.queue.shift()
+    if (item === undefined) {
+      return
+    }
+    const uri = expect(this.bundle.bundleItemToUri(item), 'item must be in bundle')
+    if (item.type === 'collections') {
+      const diagnostics = expect(await validateCollection(this.bundle, item.key), 'collection must be in bundle')
+      if (diagnostics.length === 0) {
+        return
```

Comment:
I think we need to actively percolate up / push an empty set of diagnostics for files to clear errors which are fixed. For example, currently given the following sequence of steps:

1. Introduce an error (e.g. modify image `src` to be incorrect) + save
2. Fix the error + save

Diagnostics that appear after step 1 don't go away after correction in step 2.